modules/client-handler/src/main/java/org/apache/ignite/client/handler/JdbcQueryEventHandlerImpl.java

```diff
@@ -90,62 +95,76 @@ public class JdbcQueryEventHandlerImpl implements JdbcQueryEventHandler {
      * @param meta JdbcMetadataInfo.
      * @param resources Client resources.
      */
-    public JdbcQueryEventHandlerImpl(QueryProcessor processor, JdbcMetadataCatalog meta,
-            ClientResourceRegistry resources) {
+    public JdbcQueryEventHandlerImpl(
+            QueryProcessor processor,
+            JdbcMetadataCatalog meta,
+            ClientResourceRegistry resources
+    ) {
         this.processor = processor;
         this.meta = meta;
         this.resources = resources;
     }
 
     /** {@inheritDoc} */
     @Override
-    public CompletableFuture<? extends Response> queryAsync(JdbcQueryExecuteRequest req) {
+    public CompletableFuture<JdbcConnectResult> connect() {
+        try {
+            JdbcConnectionContext connectionContext = new JdbcConnectionContext(
+                    processor::createSession,
+                    processor::closeSession
+            );
+
+            long connectionId = resources.put(new ClientResource(
+                    connectionContext,
+                    connectionContext::close
+            ));
+
+            return CompletableFuture.completedFuture(new JdbcConnectResult(connectionId));
+        } catch (IgniteInternalCheckedException exception) {
+            StringWriter sw = getWriterWithStackTrace(exception);
+
+            return CompletableFuture.completedFuture(new JdbcConnectResult(Response.STATUS_FAILED, "Unable to connect: " + sw));
+        }
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public CompletableFuture<? extends Response> queryAsync(long connectionId, JdbcQueryExecuteRequest req) {
         if (req.pageSize() <= 0) {
             return CompletableFuture.completedFuture(new JdbcQueryExecuteResult(Response.STATUS_FAILED,
                     "Invalid fetch size [fetchSize=" + req.pageSize() + ']'));
         }
 
-        QueryContext context = createQueryContext(req.getStmtType());
-
-        var results = new ArrayList<CompletableFuture<JdbcQuerySingleResult>>();
-        for (var cursorFut : processor.queryAsync(context, req.schemaName(), req.sqlQuery(),
-                req.arguments() == null ? OBJECT_EMPTY_ARRAY : req.arguments())) {
-            results.add(
-                    cursorFut.thenApply(cursor -> new JdbcQueryCursor<>(req.maxRows(), cursor))
-                            .thenCompose(cursor -> createJdbcResult(cursor, req))
-            );
+        // TODO: IGNITE-16960 support script execution
+        if (req.getStmtType() == JdbcStatementType.ANY_STATEMENT_TYPE) {
+            return CompletableFuture.completedFuture(new JdbcQueryExecuteResult(Response.STATUS_FAILED, "Scripts are not supported"));
         }
 
-        if (results.isEmpty()) {
+        JdbcConnectionContext connectionContext;
+        try {
+            connectionContext = resources.get(connectionId).get(JdbcConnectionContext.class);
```

Comment:
`resources.get(connectionId)` may return null if connection was closed concurrently/
Should we ignore request in that case?